Implement logic for creating a new incoming invoice

mladjo2505 opened the pull request:
Closes #63

mladjo2505 committed:
Implement logic for creating a new incoming invoice

stoivo commented:
- [x] Fix tabindex on invoicein.edit
- [x] Fetch period directly from fakturadato in invoicein.edit
- [x] Add new field 'unit price with tax' alongside the one we have now, unit price without tax. Unit price with tax should have higher priority than unit price without tax on invoicein.edit
- [x] Add migration for clearing 'unit price with tax' field in invoiceinline table for existing entries to 0
- [x] If we change the invoice we should update the invoice listed on invoicein.list(ex. supplier)
- [x] The amount should be calculated on save for manually created invoices
- [x] Show names for project and department on invoicein list page
- [x] Check and fix if and which JournalID the new invoice is going to get and display it correctly on the invoicein list page

mladjo2505 committed:
Fix tabindex on invoicein edit page

mladjo2505 committed:
Add unit price with tax to invoice lines in invoicein edit view

mladjo2505 committed:
Calculate total invoice amount on save for manually created invoices

mladjo2505 committed:
Rename migration to next available number

mladjo2505 committed:
Add migration for clearing unit price with tax to 0

mladjo2505 committed:
Fix invoicein import from fakturabank

Now we correctly fetch the unit cost with tax
and display the tax amount that we got when importing if any

mladjo2505 committed:
Fix calculating invoice total amount

mladjo2505 committed:
Do not show the JournalID for the invoices that cannot be bookkept

mladjo2505 committed:
Show project and department names as well on invoicein list page

mladjo2505 committed:
Show who created the invoice and when if it was not imported

mladjo2505 committed:
Update invoicein fields  correctly on date/supplier change

mladjo2505 commented on modules/fakturabank/model/fakturabank.class.php:
```diff
@@ -1443,6 +1441,8 @@ public function registerincoming() {
                         $datalineH['Vat']               = $line->TaxTotal->TaxSubtotal[0]->Percent;
                         #$datalineH['VatID']             = $line->Price->PriceAmount; #Denne mŒ nok mappes
 
+                        $datalineH['UnitCostPrice'] = $datalineH['UnitCustPrice'] + $datalineH['TaxAmount']/$Quantity;
```
TaxAmount is for Total for invoice line.
Add a comment to clarify this.
Check if Quantity can be 0.

mladjo2505 commented on modules/invoicein/model/invoicein.class.php:
```diff
@@ -179,29 +179,65 @@ function update($args) {
+        // update UnitCustPrice(unit price without tax) from UnitCostPrice(unit price with tax) if it is set
+        for($i = 1; $i <= $args['field_count']; $i++) {
+          $line_id = $args[$i];
+          if (isset($args['invoiceinline_UnitCostPrice_'.$line_id])) {
```
Change to if !isset continue.

mladjo2505 commented on modules/invoicein/model/invoicein.class.php:
```diff
@@ -179,29 +179,65 @@ function update($args) {
+          // calculate total cost for invoice
+          $TotalCustPrice = 0;
+          for($i = 1; $i <= $args['field_count']; $i++) {
+            $line_id = $args[$i];
```
Add same if from line 216 with continue statement

mladjo2505 commented on modules/invoicein/view/list.php:
```diff
@@ -180,8 +180,17 @@
       </td>
-      <td><? print $InvoiceO->Department ?></td>
-      <td><? print $InvoiceO->Project ?></td>
+      <?
```
Maybe move this to respective classes as methods instead of doing the query here just call the method.

mladjo2505 commented on modules/invoicein/view/record.inc:
```diff
@@ -21,5 +21,9 @@ if($_lib['input']->getProperty('action_invoicein_journal')) {
     print $rem->pay();
+} elseif($_lib['input']->getProperty('action_invoicein_add')) {
+    $ID = $invoicein->add();
+    header("Location: " . $_lib['sess']->dispatchs . "t=invoicein.edit&ID=" . $ID . "&inline=edit");
```
add die() here

mladjo2505 committed:
Small fixes after codereview

 Move fetching id and name for department and project to their classes
 Add some comments where needed
 Prevent devision by zero
 Prevent code propagation after header redirect

stoivo commented:
- [x] invoicein.TotalCustPrice should not be a input field since I can't save to it.
- [x] When you add a new line we get this warning: get_accountplan_object(): Mangler AccountplanID

mladjo2505 committed:
Fix: Just print invoicein total amount, resolve missing accountplan

mladjo2505 commented on modules/invoicein/model/invoicein.class.php:
```diff
@@ -179,29 +179,65 @@ function update($args) {
+          if(!$accountplan->City) $_lib['message']->add('Sted mangler p&aring; leverand&oslash;r kontoplan');
+          $args['invoicein_IName_' . $ID] = $accountplan->AccountName;
+          if(!$accountplan->AccountName) $_lib['message']->add('Navn mangler p&aring; leverand&oslash;r kontoplan');
+          $args['invoicein_IAddress_' . $ID] = $accountplan->Address;
```
Check out why this shows up on initial create for invoice

mladjo2505 commented:
- [x] On create new invoicein page do not show 'Er med i reisegarantifondet'

mladjo2505 commented:
On invoiceinline if we choose a car enabled accountplan and after choose another one with car disabled, clear car from that line? also globally? create a new issue?

mladjo2505 commented:
- [x] Add an indicator in invoicein list page which invoices were imported and which were manually created.

mladjo2505 commented:
- [x] Bilagsnumber should be shown as a link to a journal if the invoice is bookkept and empty if not, neither the imported nor the manually created invoices should have an input field which allows us to change it.

mladjo2505 commented:
- [x] Comment out remitering stuff (both on invoice in and from menu) for now. See what to do with it later.

mladjo2505 commented:
- [x] Add delete logic for manually created invoicein and only allow it for not yet bookkept invoices.

mladjo2505 commented:
- [x] When we change back to an account without a car from an account with a car remove the car from the line/row.

mladjo2505 commented:
@stoivo I see you added the "Back to rework" label to this. Other than the commented things to be done, are there any more requirements?

stoivo commented:
@hildegunn95 will test it again when that is done. 
But I don't have any other things to say.

mladjo2505 committed:
Fix after code review and discussion

mladjo2505 commented:
Fixed all requests and rebased to new master.

mladjo2505 commented on modules/invoicein/view/edit.php:
```diff
@@ -126,8 +121,13 @@
         }
         ?>
         </td>
+<? if (!$invoicein->Imported) { ?>
```
When exactly should this be shown and when not?
@stoivo

mladjo2505 commented on modules/invoicein/view/edit.php:
```diff
@@ -308,8 +312,19 @@
+          }
         ?>
+      </td>
+    </tr>
```
Access level and rights who can create invoicein and delete?
@stoivo

sls-lnluksa commented:
First round review: Nikola Lukovic (sls-lnluksa)
Final review with: Nikola Lukovic, @stoivo 

There are two questions that @stoivo needs to answer to finish this review.

mladjo2505 commented:
Fixed one typo and squashed it to the last commit.

sls-lnluksa commented:
@mladjo2505 
1. Access level and rights who can create invoicein and delete?
Same as on vouchers.
2. On create new invoicein page do not show 'Er med i reisegarantifondet'
After discussing it with @stoivo, we decided to remove it from all invoicein. Both manually added and imported.

mladjo2505 commented:
1. Same as on vouchers if the period is open and you have the sufficient access.
2. Remove 'Er med i reisegarantifondet'
3. get_accountplan_object(): Mangler AccountplanID should not be showing, see why.

mladjo2505 committed:
Small fixes after code review

mladjo2505 commented:
Went through latest changes with @MrdjanStojimirovic 
Final review with @stoivo

mladjo2505 commented:
Rebased to new master

mladjo2505 commented:
Issues and things to look at after initial testing with @hildegunn95 
- [x] Date validation for invoice date and period (something similar to #225 )
- [x] When bookkeeping no invoices it prints out just `fakturaer er bilagsført` and no number, figure out why and add 0 there if that is the case
- [x] Betaling column on invoicein.list page. What it is (EDIT: it is PaymentMeansCode) and why it is printed as a number for example 10, 48, 42, 0 (EDIT: Now prints text instead of code)
- [x] Quantity column on invoice line should be formatted as amount and shown with 3 decimal places
- [x] get_accountplan_object() still showing `Mangler AccountPlanID`, find out why? (EDIT: Caused by starting for loop from 0 instead of 1. Now fixed)
- [x] When bookkeeping it prints `bilagsfør` at the top of the page, find out why and remove
- [x] Add a warning message when no supplier is selected (Leverandør ikke valgt!).
- [x] Check if the Status column on invoicein.list page is an indicator that an invoice is ready to be bookkept if it is empty.
- [x] Total invoice amount is 0 if an empty invoice line is present, it remains so even after we delete that line until we save the invoice without it (EDIT: Now adding and deleting lines recalculates the invoice total amount correctly).
- [x] Calculate due date from invoice date and credit days setting
- [x] Showing `Mangler adresse på leverandør...` but the address is present on supplier
- [x] Check if there is any more remitterings stuff to remove

Fix typos on invoicein.edit page:
- [x] Org nr -> Orgnr
- [x] Faktura dato -> Fakturadato
- [x] Forfalls dato -> Forfallsdato
- [x] Faktura periode -> Fakturaperiode
- [x] Leverings betingelse -> Leveringsbetingelse
- [x] Betalings betingelse -> Betalingsbetingelse
- [x] Opprettet på -> Opprettet
- [x] Produkt navn -> Produktnavn
- [x] Enhetspris inklusiv MVA -> Enhetspris inkl. MVA
- [x] Beløp U/MVA -> Beløp eksl. MVA

mladjo2505 committed:
Fix typos on invoicein.edit page

mladjo2505 committed:
Fix more stuff

mladjo2505 committed:
Add invoice date validation to incoming invoice

mladjo2505 commented:
Rebased to new master

mladjo2505 committed:
Payment means should be printed as text

mladjo2505 committed:
Fix for missing accountplan id for invoice in lines

mladjo2505 commented:
Should we add some validations on incoming invoices and allow them to be bookkept only after they pass those validations? (EDIT: Just add status shown on list page to invoice edit page as a warning message)

Should we remove the following since they are all connected to remitterings: 
- Status input field in incoming invoices search form(RemittanceStatus)
- Remittering column from invoice list table

mladjo2505 committed:
Recalculate invoice on save and delete
 Fix missing address on supplier
 Fix missing AccountPlanID

stoivo commented:
We should have the same validations and we have on the list page.
We should remove Status from the search form and not print the RemittanceStatus. Still keep the db fields.

mladjo2505 committed:
Add warning errors to invoice incoming

mladjo2505 committed:
Remove more remittance stuff

mladjo2505 commented:
- [x] Prediction of JournalID for invoices that are ok for bookkeeping is not working well (if there is a gap between two it will take them into account as they are to be bookkept even if they have errors)
- [x] Add a link with `'-'` as text for it if no invoice number is set
- [x] If the invoice is created without an invoice date, use the current date set for this session on the top of page form.

mladjo2505 commented on db/changes/127_imported_to_invoicein_table.sql:
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE invoicein
+ADD Imported int(1) DEFAULT 1;
+
+UPDATE invoiceinline SET UnitCostPrice = 0 WHERE Active = 1;
```
Check this one more time with @stoivo

mladjo2505 commented on modules/department/model/department.class.php:
```diff
@@ -0,0 +1,19 @@
+
+  public function getDepartmentIDAndName() {
+    global $_lib;
+    $query_department = sprintf("SELECT * FROM companydepartment WHERE CompanyDepartmentID = %d", $this->DepartmentID);
```
Add a check if the `$this->DepartmentID` is set.

mladjo2505 commented on modules/fakturabank/model/fakturabank.class.php:
```diff
@@ -1426,23 +1426,31 @@ public function registerincoming() {
                         $datalineH['UnitCostPriceCurrencyID'] = exchange::getLocalCurrency();
                         $datalineH['UnitCustPriceCurrencyID'] = exchange::getLocalCurrency();
 
+                        // TaxAmount fields here refers to the Total tax amount
```
field*

mladjo2505 commented on modules/invoicein/model/invoicein.class.php:
```diff
@@ -88,6 +88,9 @@ function fill($args) {
         }
+        if($this->ID) {
+            $query .= " i.ID = '$this->ID' and ";
+        }
```
possibly not needed

mladjo2505 commented on modules/invoicein/model/invoicein.class.php:
```diff
@@ -176,39 +179,98 @@ function update($args) {
+        }
+        if(!$invoicein->DAddress) {
+                $args['invoicein_DAddress_' . $ID] = $accountplan->Address;
+        }
```
All the delivery address fields should go to the upper if block, since they should be updated on account plan change.

mladjo2505 commented on modules/invoicein/model/invoicein.class.php:
```diff
@@ -176,39 +179,98 @@ function update($args) {
+        if(!$invoicein->CustomerAccountPlanID) {
+          $args['invoicein_CustomerAccountPlanID_' . $ID] =  $_lib['sess']->get_companydef('OrgNumber');
+        }
+        $args['invoicein_RemittanceAmount_'.$ID] = $args['invoicein_TotalCustPrice_'.$ID];
```
Remove remittance stuff.

mladjo2505 commented on modules/invoicein/view/edit.php:
```diff
@@ -202,8 +227,16 @@
+      $vatline = $row2->TaxAmount;
+    } else {
+      $vatline = ($row2->Vat/100) * $sumline;
+    }
```
The above if could be changed to recalculate if it is a manually created invoice. Discussion needed.
@stoivo & @hildegunn95 You should see this and think about it.

mladjo2505 commented on modules/invoicein/view/record.inc:
```diff
@@ -1,17 +1,26 @@
-    $invoicein->update($_POST);
     $invoicein->linedelete($_REQUEST);
 
 } elseif($_lib['input']->getProperty('action_invoicein_remittance')) {
```
remove

mladjo2505 commented on modules/project/model/project.class.php:
```diff
@@ -0,0 +1,19 @@
+
+  public function getProjectIDAndName() {
+    global $_lib;
+    $query_project = sprintf("SELECT * FROM project WHERE ProjectID = %d", $this->ProjectID);
```
Check if `$this->ProjectID` is set.

mladjo2505 committed:
Additional fixes after code review

mladjo2505 commented:
Unresolved things on this PR:
- [x] `UPDATE invoiceinline SET UnitCostPrice = 0 WHERE Active = 1;` line in the migration. Discuss with @stoivo 
- [x] On imported invoices TaxAmount saved is only displayed, but if we create a new manual invoice that field is never saved to the database. And this causes the lines to have the same TaxAmount even after they are changed since the one loaded from the database is printed. Discuss with @stoivo & @hildegunn95 
**EDIT**: Tested this out and it seems to recalculate after rebase.

mladjo2505 commented:
We should rebase this after EHF is merged into master and resolve all conflicts, then test this thoroughly.

mladjo2505 commented:
Rebased this to `allowances_and_charges` branch.
Should be further rebased to new master after #277 is merged in master.
Should be tested more if it still retains the logic requested.

mladjo2505 committed:
Fix invoicein show after rebase

mladjo2505 commented:
- [x] Does not work well with allowance/charge, Invoice total price in header table is not calculated with a/c. (Due to a recalculating function that was only on this branch)
- [x] Allowance/Charge for lines has fewer columns and are not aligned with invoice line.

mladjo2505 committed:
Fix recalculate on save an some alignment on edit page

mladjo2505 commented:
Fixed, and rebased to newest master.

mladjo2505 commented:
- [x] Reorder buttons to look the same as on outgoing invoice page (change: save button should be to the left, and delete to the right. Like creating outgoing invoices)
- [x] Rename Enhetspris -> Enhetspris ekskl. MVA(make sure to break inkl/ekskl together with MVA)
- [x] When you save without supplier, page turns white (EDIT: Now it shows an error message that the supplier is not selected)
- [x] When you save the invoice in a closed period, the save button disappears -> Can't save to a locked period(it should not be possible) (EDIT: Now shows an error message that you tried to save to a locked period)
- [x] Why is period selectable when it always saves the same period as the issue date? Change from dropdown to printed text (EDIT: Done in #444 When that PR is merged, we can rebase this)
- [x] If you enter the same invoice number and supplier as an invoice already punched => bad update and ugly print (EDIT: Now it shows an error message that the invoice with that supplier and invoice number is already punched)
- [x] Print the warning ‘En eller flere fakturalinjer mangler resultatkonto’ on a new line, for instance no result account, and invoice amount is 0

sls-lnluksa commented:
- [x] It is not possible to enter invoicein_edit page, therefor it is not possible to test page.
`Ny innkommende faktura` is not working.
- [x] Remove warning 'Leverandør ikke valgt' from invoicein.list

mladjo2505 committed:
Renumber migration

mladjo2505 commented:
If any more changes are to be made to this it should be squashed into one commit for convenience of rebasing.
Tested on staging lodo2016:

- Creating new incoming invoice
- Adding a new line
- Adding an allowance/charge
- Adding a line allowance/charge
- Removing a line
- Removing an allowance/charge
- Removing a line allowance/charge

sls-lnluksa commented:
1. round review -> needed 1h.

mladjo2505 commented:
- [x] Rebase and fix (1h)
